slick/src/main/scala/slick/lifted/Query.scala

```diff
@@ -61,7 +61,7 @@ sealed abstract class Query[+E, U, C[_]] extends QueryBase[C[U]] { self =>
 
   /** Applies the given filter function, if the boolean parameter `p` evaluates to true. 
     * If not, the filter will not be part of the query. */
-  def filterIf(p: Boolean)(f: E => Rep[Boolean]): Query[E, U, C] =
+  def filterIf[T <: Rep[_] : CanBeQueryCondition](p: Boolean)(f: E => T): Query[E, U, C] =
```

Comment:
Is the `<: Rep[_]` necessary? Or could it just be

```suggestion
  def filterIf[T : CanBeQueryCondition](p: Boolean)(f: E => T): Query[E, U, C] =
```

Reply:
Hmm, good point @nafg. `CanBeQueryCondition` is already a function from its parameter type to a `Rep`. Looking at the call chain `filterIf` => `withFilter` => `filterHelper`, the last two don't need `T < Rep[?]`, so I think we can remove the type bounds from other functions as well. Here are the current definitions:
```scala
trait CanBeQueryCondition[-T] extends (T => Rep[_])

...

  private def filterHelper[T](f: E => T, wrapExpr: Node => Node)
                             (implicit wt: CanBeQueryCondition[T]): Query[E, U, C] = {
    val generator = new AnonSymbol
    val aliased = shaped.encodeRef(Ref(generator))
    val fv = f(aliased.value)
    new WrappingQuery[E, U, C](Filter.ifRefutable(generator, toNode, wrapExpr(wt(fv).toNode)), shaped)
  }

  def filter[T <: Rep[_]](f: E => T)(implicit wt: CanBeQueryCondition[T]): Query[E, U, C] =
    withFilter(f)

  def filterNot[T <: Rep[_]](f: E => T)(implicit wt: CanBeQueryCondition[T]): Query[E, U, C] =
    filterHelper(f, node => Library.Not.typed(node.nodeType, node) )

  def filterOpt[V, T <: Rep[_] : CanBeQueryCondition](optValue: Option[V])(f: (E, V) => T): Query[E, U, C] =
    optValue.map(v => withFilter(a => f(a, v))).getOrElse(this)

  def filterIf[T <: Rep[_] : CanBeQueryCondition](p: Boolean)(f: E => T): Query[E, U, C] =
    if (p) withFilter(f) else this

  def withFilter[T : CanBeQueryCondition](f: E => T) = filterHelper(f, identity)

```

That being said, we can also remove `T < Rep[?]` bound from not only `filterIf` but also `filter`, `filterNot` and `filterOpt`, too.

```scala
  def filter[T](f: E => T)(implicit wt: CanBeQueryCondition[T]): Query[E, U, C] =
    withFilter(f)

  def filterNot[T](f: E => T)(implicit wt: CanBeQueryCondition[T]): Query[E, U, C] =
    filterHelper(f, node => Library.Not.typed(node.nodeType, node) )

  def filterOpt[V, T : CanBeQueryCondition](optValue: Option[V])(f: (E, V) => T): Query[E, U, C] =
    optValue.map(v => withFilter(a => f(a, v))).getOrElse(this)

  def filterIf[T : CanBeQueryCondition](p: Boolean)(f: E => T): Query[E, U, C] =
    if (p) withFilter(f) else this
```

Reply:
Made the suggested changes 👍